### PR TITLE
Candidate for 4.0.41

### PR DESCRIPTION
**Pull request by kushti**
* #1599 
* #1786
* #1812 
* #1814 

Contributors: @ApexTheory , @knizhnik , @kushti

### CODE REVIEW

File: src/main/scala/org/ergoplatform/modifiers/mempool/UnconfirmedTransaction.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+object UnconfirmedTransactionSerializer extends ScorexSerializer[UnconfirmedTransaction] {
+
+  override def serialize(unconfirmedTx: UnconfirmedTransaction, w: Writer): Unit = {
+    ErgoTransactionSerializer.serialize(unconfirmedTx.transaction, w)
+    w.putOption(unconfirmedTx.lastCost)(_.putUInt(_))
+    w.putULong(unconfirmedTx.createdTime)
+    w.putULong(unconfirmedTx.lastCheckedTime)
+    w.putOption(unconfirmedTx.transactionBytes){(_,d) =>
+      w.putInt(d.length)
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/mempool/UnconfirmedTransaction.scala:48]: better use unsigned encoding of Int (putUInt), it is more efficient

File: src/main/scala/org/ergoplatform/modifiers/mempool/UnconfirmedTransaction.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  // this serializer used internally only, so we do not validate data here
+  override def parse(r: Reader): UnconfirmedTransaction = {
+    val ergoTransaction = ErgoTransactionSerializer.parse(r)
+    val lastCostOpt = r.getOption(r.getUInt().toInt)
+    val createdTime = r.getULong()
+    val lastCheckedTime = r.getULong()
+    val transactionBytesOpt = r.getOption{
+      val size = r.getInt()
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/mempool/UnconfirmedTransaction.scala:61]: getUInt() (see above comment)

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: src/main/scala/org/ergoplatform/http/api/ApiCodecs.scala [hunk 3]
```diff
@@ -195,6 +195,27 @@ trait ApiCodecs extends JsonCodecs {
     } yield ErgoTransaction(ergoLikeTx)
   }
 
+  // We are not using this encoder for now, but may use in future
+  implicit val unconfirmedTxEncoder: Encoder[UnconfirmedTransaction] = { unconfirmedTx =>
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/http/api/ApiCodecs.scala:199]: add restoring bytes from tx

File: src/main/scala/org/ergoplatform/modifiers/mempool/UnconfirmedTransaction.scala (hunk 1 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/mempool/UnconfirmedTransaction.scala:48]: removed unused serializer

File: src/main/scala/org/ergoplatform/modifiers/mempool/UnconfirmedTransaction.scala (hunk 2 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/mempool/UnconfirmedTransaction.scala:61]: see above

File: src/main/scala/org/ergoplatform/http/api/ApiCodecs.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/http/api/ApiCodecs.scala:199]: no need to store bytes as they can be derived from tx

**Review by aslesarenko — APPROVED** [review]

**Pull request merged**